### ISSUE TITLE
[dart2wasm] Assign stack traces to `Error` objects on `throw`

### ISSUE THREAD

**Issue by osa1** [dart-lang/sdk#50608]
Language spec says this about throwing an error: (section 17.10)

> If v is an instance of class Error or a subclass thereof, and it is the first time that Error object is thrown, the stack trace t is stored on v so that it will be returned by the stackTrace getter inherited from Error.

For this we need to have a field in `Error` class to store the stack trace, and update it in `throw`.

I implemented this in https://dart-review.googlesource.com/c/sdk/+/273142. VM adds a `_stackTrace` field to `Error` which we copy in dart2wasm patch file. However it broke a few tests. The problem is, given an `Error` object, we have no way to say in runtime whether it "extends" `Error` or "implements" it. If it extends `Error`, then it will have the field and we should update it. If it implements `Error`, then it won't have the field and we cannot update it.

(The spec says the `stackTrce` getter inherited from `Error` should return the call stack, and since with "implements" the class won't be inheriting it, it's not a violation of the spec to not assign a stack trace to the `Error` object if it "implements" `Error`)

So currently we have no way of implementing this part of the spec, we need extra information in runtime.

Some ideas:

- We could generate an array with class IDs of `Error` subclasses. In `throw`, check if thrown object's class ID is in the array, and set the field if it is.
- Similar to the above, but we number classes so that `Error` subclasses get consecutive class IDs. The check then becomes a simple range check.

**Comment by lrhn** [comment]
You probably need to do something at compile time, if you don't have runtime information.

If all else fails, maybe you can store it on the objects, as synthetic members?

E.g., add an `bool get _extendsError => true;` getter to `Error`, and in the compiler introduce a synthetic `_bool get _extendsError => false;` on every class which implements `_Error`, but doesn't extend it. That's any class which needs an implementation of that particular method, but doesn't inherit one. Those classes would normally be given a `noSuchMethod`-thrower implementation. You "just" need to recognize that and change it to a working version.

(The consecutive class IDs idea is definitely better if you can make it work. Maybe _always_ order class IDs consecutively for extending classes. The super-class hierarchy is a tree, so it should be possible. Might also be useful for other things.)

**Comment by osa1** [comment]
We briefly discussed this in chat. Adding a getter should be simple enough. In runtime there will be two functions, one that returns `true`, one that returns `false`, and those functions will be added to the dispatch table for the `_extendsError` calls (one entry per `Error` subtype).

I think currently selectors correspond to Dart members (`DispatchTable` builds selectors from members), but for this feature we will need a selector without actual Dart members. That may require some refactoring, but it shouldn't be too difficult.

**Comment by osa1** [comment]
> I think currently selectors correspond to Dart members (DispatchTable builds selectors from members), but for this feature we will need a selector without actual Dart members. That may require some refactoring, but it shouldn't be too difficult.

I tried doing this in #50014 and it was a hacky mess. I think a simpler approach would be to add synthetic members to kernel `Class`es in a kernel transform.

**Comment by osa1** [comment]
> I think a simpler approach would be to add synthetic members to kernel Classes in a kernel transform.

I tried this in a branch: https://github.com/dart-lang/sdk/compare/main...osa1:sdk:dart2wasm_error_stack_traces?expand=1 the transformation is simple, but the procedures added in the transformation are lost before code generation. Adding `@pragma('wasm:entry-point')` to the new procedures doesn't make a difference.

I also realized that even if the procedures survived the front-end passes the dispatch table builder wouldn't generate entries for these members, because they are not part of an interface and won't have the same selector IDs (`ProcedureAttributesMetadata.methodOrSetterSelectorId`s will be different). I think there is probably some way of making these `_extendsError`s part of the `Error` interface, but I'm not sure how.

I think generalizing dispatch table builder to allow synthetic members may actually be easier that I thought, I'll give this a try next.

**Comment by osa1** [comment]
> I think generalizing dispatch table builder to allow synthetic members may actually be easier that I thought, I'll give this a try next.

I did this in https://github.com/osa1/sdk/commit/5ede69f92220936639d6a1819ad667ce9a1c0b12 (warning: hacky code).

It works fine, but I realized while testing this that it's possible to throw any type of object, not just `Error`.

What that means is, when checking whether we can update the `Error._stackTrace` field, it's not possible to just call `error._extendsError` because `error` may not even be an `Error` subtype.

So we can't avoid the runtime type check in `throw`, and if we're going to do a costly runtime type check, we could instead have a runtime function specifically for checking if a class ID is for a class that "extends" the `Error` class.

As far as I can see, `_TypeUniverse` currently cannot distinguish "extends" from "implements". So we can't reuse it.

We can generate an array of class IDs for classes that extend `Error` and then a function to do binary search on it.

I'm not quite sure how to implement this search function though. As far as I understand from the GC spec, it's not possible to allocate a static Wasm array. We could create a data segment, but then we would have to implement the search function in Wasm as we can't read data segments from Dart.

For now, I'm thinking of adding a new `Error` member:

```dart
@patch
class Error {
  external static bool _extendsError(int classID);
}
```

and implement it as an if-else chain to compare the classID with the classIDs of classes known to extend `Error`. It will be slower than binary search, but I wouldn't be surprised if it's actually faster than `_isSubtype`.
